nodes/xrd/xrd.go

```diff
@@ -0,0 +1,139 @@
+		"XR_FIRST_BOOT_CONFIG": "/etc/xrd/first-boot.cfg",
+		"XR_MGMT_INTERFACES":   "linux:eth0,xr_name=Mg0/RP0/CPU0/0,chksum,snoop_v4,snoop_v6",
+	}
+
+	//go:embed xrd.cfg
+	cfgTemplate string
+)
+
+const (
+	scrapliPlatformName = "cisco_iosxr"
+	defaultUser         = "clab"
+	defaultPassword     = "clab@123"
+)
+
+// Register registers the node in the global Node map.
+func Register() {
+	nodes.Register(kindnames, func() nodes.Node {
+		return new(xrd)
+	})
+	err := nodes.SetDefaultCredentials(kindnames, defaultUser, defaultPassword)
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+type xrd struct {
+	nodes.DefaultNode
+}
+
+func (n *xrd) Init(cfg *types.NodeConfig, opts ...nodes.NodeOption) error {
+	// Init DefaultNode
+	n.DefaultNode = *nodes.NewDefaultNode(n)
+
+	n.Cfg = cfg
+	for _, o := range opts {
+		o(n)
+	}
+
+	n.Cfg.Binds = append(n.Cfg.Binds,
+		// mount first-boot config file
+		fmt.Sprint(filepath.Join(n.Cfg.LabDir, "first-boot.cfg"), ":/etc/xrd/first-boot.cfg"),
+		// persist data by mounting /xr-storage
+		fmt.Sprint(filepath.Join(n.Cfg.LabDir, "xr-storage"), ":/xr-storage"),
+	)
+
+	return nil
+}
+
+func (n *xrd) PreDeploy(ctx context.Context, _, _, _ string) error {
+	n.genInterfacesEnv()
+
+	utils.CreateDirectory(n.Cfg.LabDir, 0777)
+
+	return n.createXRDFiles(ctx)
+}
+
+func (n *xrd) SaveConfig(_ context.Context) error {
+	err := netconf.SaveConfig(n.Cfg.LongName,
+		defaultUser,
+		defaultPassword,
+		scrapliPlatformName,
+	)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("saved %s running configuration to startup configuration file\n", n.Cfg.ShortName)
+	return nil
+}
+
+func (n *xrd) createXRDFiles(_ context.Context) error {
+	nodeCfg := n.Config()
+	// generate xr-storage directory
+	utils.CreateDirectory(filepath.Join(n.Cfg.LabDir, "xr-storage"), 0777)
+	// generate first-boot config
+	cfg := filepath.Join(n.Cfg.LabDir, "first-boot.cfg")
+	nodeCfg.ResStartupConfig = cfg
+
+	// set mgmt IPv4/IPv6 gateway as it is already known by now
+	// since the container network has been created before we launch nodes
+	// and mgmt gateway can be used in xrd.Cfg template to configure default route for mgmt
+	nodeCfg.MgmtIPv4Gateway = n.Runtime.Mgmt().IPv4Gw
+	nodeCfg.MgmtIPv6Gateway = n.Runtime.Mgmt().IPv6Gw
+
+	// use startup config file provided by a user
+	if nodeCfg.StartupConfig != "" {
+		c, err := os.ReadFile(nodeCfg.StartupConfig)
+		if err != nil {
+			return err
+		}
+		cfgTemplate = string(c)
+	}
+
+	err := n.GenerateConfig(nodeCfg.ResStartupConfig, cfgTemplate)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// genInterfacesEnv calculates how many interfaces were defined for xrd node
+// and populates the content of a required env var.
+func (n *xrd) genInterfacesEnv() {
+	// xrd (at least control-plane variant) needs XR_INTERFACE ENV var to be populated for all active interface
+	// here we take the number of links users set in the topology to get the right # of links
+	var interfaceEnvVar string
+
```

Comment:
Should this not maybe reference the names from the endpoint config, rather then the index var I?!
This will lead to confusion, not?!

Reply:
what do you mean, @steiler?
the `i` is the index of the endpoints slice element

Reply:
interfaceEnvVar += fmt.Sprintf("linux:eth%d,xr_name=Gi0/0/0/%d;", i+1, i)
Yes, and then the index is used but not the data behind the index...
Or am I missing something here myself?!

Reply:
@steiler I think you refer to the usage of `linux:eth%d` whereas it could've been the entire endpoint name listed here instead of an index?
Yes, I think I can do that, to make sure that users might say

```
["xrd:eth5", "srl:e1-1"]
```

and this still would work

Reply:
@steiler like this [ecdf202](https://github.com/srl-labs/containerlab/pull/1144/commits/ecdf2021ef2f7b5dda32bbfc48ad25c9fb1d6022)

Reply:
Well yes, in parts...
Would you not also want to allow interface numbers on the xrd side that is not consecutive...
I'm thinking someone who tries to mimic the prod network... Has connections on gi0/0/0/24 and does not want to change interface names in the config ... So can't we come up wit a schema that also allows non consecutive xrd interface numbers?
Wouldn't that make sense?!
We also do so in the srl case.

Reply:
@steiler makes sense. Let me try that

Reply:
@steiler check [e262674](https://github.com/srl-labs/containerlab/pull/1144/commits/e2626748ce30bb35f6b476192fa22f161d956fc6). Sounds about right?

Reply:
Yes, I'd say that's in line with my thinking.
Thanks!

Reply:
@steiler nice, thanks. I completely missed on this opportunity